javascript/src/index.js

```diff
@@ -72,160 +73,356 @@ function preventUnintendedZoomOnScroll(map) {
-          data.limits.lat[1] = data.limits.lat[1] + pad;
+        // Returns true if the zoomToLimits option says that the map should be
+        // zoomed to map elements.
+        function needsZoom() {
+          return options.zoomToLimits === "always" ||
+                 (options.zoomToLimits === "first" && !map.leafletr.hasRendered);
         }
-        if (data.limits.lng[0] === data.limits.lng[1]) {
-          data.limits.lng[0] = data.limits.lng[0] - pad;
-          data.limits.lng[1] = data.limits.lng[1] + pad;
+
+        if (!explicitView && needsZoom() && !map.getZoom()) {
+          if (data.limits) {
+            // Use the natural limits of what's being drawn on the map
+            // If the size of the bounding box is 0, leaflet gets all weird
+            let pad = 0.006;
+            if (data.limits.lat[0] === data.limits.lat[1]) {
+              data.limits.lat[0] = data.limits.lat[0] - pad;
+              data.limits.lat[1] = data.limits.lat[1] + pad;
+            }
+            if (data.limits.lng[0] === data.limits.lng[1]) {
+              data.limits.lng[0] = data.limits.lng[0] - pad;
+              data.limits.lng[1] = data.limits.lng[1] + pad;
+            }
+            map.fitBounds([
+              [ data.limits.lat[0], data.limits.lng[0] ],
+              [ data.limits.lat[1], data.limits.lng[1] ]
+            ]);
+          } else {
+            map.fitWorld();
+          }
         }
-        map.fitBounds([
-          [ data.limits.lat[0], data.limits.lng[0] ],
-          [ data.limits.lat[1], data.limits.lng[1] ]
-        ]);
-      } else {
-        map.fitWorld();
-      }
-    }
 
-    for (let i = 0; data.calls && i < data.calls.length; i++) {
-      let call = data.calls[i];
-      if (methods[call.method])
-        methods[call.method].apply(map, call.args);
-      else
-        log("Unknown method " + call.method);
-    }
+        for (let i = 0; data.calls && i < data.calls.length; i++) {
+          let call = data.calls[i];
+          if (methods[call.method])
+            methods[call.method].apply(map, call.args);
+          else
+            log("Unknown method " + call.method);
+        }
 
-    map.leafletr.hasRendered = true;
+        map.leafletr.hasRendered = true;
 
-    if (!HTMLWidgets.shinyMode) return;
+        if (HTMLWidgets.shinyMode){
+          setTimeout(function() { updateBounds(map); }, 1);
+        }
 
-    setTimeout(function() { updateBounds(map); }, 1);
-  },
-  resize: function(el, width, height, map) {
-    map.invalidateSize();
-    if (map.leafletr.pendingRenderData) {
-      this.doRenderValue(el, map.leafletr.pendingRenderData, map);
-    }
+      },
+      resize: function(width, height) {
+        if(map) {
+          map.invalidateSize();
+          if (map.leafletr.pendingRenderData) {
+            this.doRenderValue(map.leafletr.pendingRenderData, map);
+          }
+        }
+      }
+    };
   }
 });
 
+//###HTMLWidgets.widget({
```

Comment:
Yup, I left it for the code review part, coz git diff was wanky for this portion, so easier to check new code against commented code in same file. :)